#! /bin/sh /usr/share/dpatch/dpatch-run
## 57_fix_channel_number_select_for_direct_plugins.dpatch by Daniel T Chen <crimsun@ubuntu.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix channel number selection for PCM direct plugins
## DP: http://hg-mirror.alsa-project.org/alsa-lib?cs=6f31cef294b6;style=raw

@DPATCH@
diff -urNad alsa-lib-1.0.13~/src/pcm/pcm_direct.c alsa-lib-1.0.13/src/pcm/pcm_direct.c
--- alsa-lib-1.0.13~/src/pcm/pcm_direct.c	2007-02-04 19:39:18.000000000 -0500
+++ alsa-lib-1.0.13/src/pcm/pcm_direct.c	2007-02-04 19:39:56.000000000 -0500
@@ -899,28 +899,10 @@
 		}
 		params->format = format;
 	}
-	ret = snd_pcm_hw_params_set_channels(spcm, hw_params, params->channels);
+	ret = INTERNAL(snd_pcm_hw_params_set_channels_near)(spcm, hw_params, (unsigned int *)&params->channels);
 	if (ret < 0) {
-		unsigned int min, max;
-		ret = INTERNAL(snd_pcm_hw_params_get_channels_min)(hw_params, &min);
-		if (ret < 0) {
-			SNDERR("cannot obtain minimal count of channels");
-			return ret;
-		}
-		ret = INTERNAL(snd_pcm_hw_params_get_channels_min)(hw_params, &max);
-		if (ret < 0) {
-			SNDERR("cannot obtain maximal count of channels");
-			return ret;
-		}
-		if (min == max) {
-			ret = snd_pcm_hw_params_set_channels(spcm, hw_params, min);
-			if (ret >= 0)
-				params->channels = min;
-		}
-		if (ret < 0) {
-			SNDERR("requested count of channels is not available");
-			return ret;
-		}
+		SNDERR("requested count of channels is not available");
+		return ret;
 	}
 	ret = INTERNAL(snd_pcm_hw_params_set_rate_near)(spcm, hw_params, (unsigned int *)&params->rate, 0);
 	if (ret < 0) {
